File: python/subproblem_setups/sp6.py
#Four cones
from linearSubproblemSltns import sp6_lib as sp6
import time as time

#Helper variables
H = sp6.np.zeros((4, 3))
K = sp6.np.zeros((4, 3))
P = sp6.np.zeros((4, 3))

#Begin solving
numTests = input("Enter the number of tests: ")

#Input formatting
numTests.strip()
numTests = int(numTests) #Convert from str to int

timeSum = 0  #Keep track of time for all operations
errorSum = 0 #Keep track of errors
time1 = 0    #Start time
time2 = 0    #End time

#Run procedure
for i in range(numTests):
   #Perform setup
   d1, d2 = sp6.sp6_setup(H, K, P)
   time1 = time.perf_counter_ns()

   #Perform calculations
   theta1, theta2 = sp6.sp6_run(H, K, P, d1, d2)
   time2 = time.perf_counter_ns()
   timeSum += time2-time1 #Collect time data
   errorSum += sp6.sp6_error(H, K, P, theta1, theta2, d1, d2)

#Print results
print("\r\nResults: (Number of Tests: {})\r\n".format(numTests) + "-"*50)
print("Avg Time: {} ns, Total Time {} ns".format(timeSum/numTests, timeSum))
print("Average error margin: {}".format(errorSum/numTests))
print("Time in s: {}\r\n".format(timeSum/1000000000))

File: matlab/+subproblem_setups/sp_6.m
classdef sp_6
    methods (Static)
        function [P, S] = setup()
            P.H = rand_normal_vec(4);
            P.K = rand_normal_vec(4);
            P.P = rand_vec(4);

            
            S.theta1 = rand_angle;
            S.theta2 = rand_angle;
            
            P.d1 =  P.H(:,1)'* rot(P.K(:,1), S.theta1) * P.P(:,1) ...
                  + P.H(:,2)'* rot(P.K(:,2), S.theta2) * P.P(:,2);
            P.d2 =  P.H(:,3)'* rot(P.K(:,3), S.theta1) * P.P(:,3) ...
                  + P.H(:,4)'* rot(P.K(:,4), S.theta2) * P.P(:,4);
        end

        function S = run(P)
        [S.theta1, S.theta2] = subproblem.sp_6(P.H, P.K, P.P, P.d1, P.d2);
        end

        function e = error(P, S)  
            for i = 1:length(S.theta1)
                e_i = norm([P.H(:,1)'* rot(P.K(:,1), S.theta1(i)) * P.P(:,1) ...
                          + P.H(:,2)'* rot(P.K(:,2), S.theta2(i)) * P.P(:,2) - P.d1
                            P.H(:,3)'* rot(P.K(:,3), S.theta1(i)) * P.P(:,3) ...
                          + P.H(:,4)'* rot(P.K(:,4), S.theta2(i)) * P.P(:,4) - P.d2]);
                if i == 1
                    e = e_i;
                else
                    e = e + e_i;
                end
            end
        end
    end
end